proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+    // The training is running without error.
+    TFJobRunning TFJobPhase = "Running"
+
+    // All pods in the TFJob have terminated in success,
+    // and will not be restarted.
+    // The training is complete without error.
+    TFJobSucceeded TFJobPhase = "Succeeded"
+
+    // All pods in the TFJob have terminated, and at least one pod
+    // has terminated in failure.
+    // The training is failed.
+    TFJobFailed TFJobPhase = "Failed"
+
+    // For some reason the state of the TFJob could not be obtained,
+    // typically due to an error in communicating with the host of the TFJob.
+    TFJobUnknown TFJobPhase = "Unknown"
+)
+```
+
+## Design
+
+### TFController
+
+The TFJob controller `tf-operator` will process TFJobs and CRUD services/pods according to the spec of TFJob. It is responsible for synchronizing TFJob objects stored in the system with actual running services and pods, continuously strive to make the observed state match the desired state.
+
+Here is the definition of `TFController`:
+
+```go
+type TFController struct {
+    // kubeClientset is a standard kubernetes clientset
+    kubeClientset kubernetes.Interface
+
+    // tfJobClientset is a clientset for CRD TFJob
+    tfJobClientset tfjobclient.Interface
+
+    tfJobLister listers.TFJobLister
+    tfJobSynced cache.InformerSynced
+
+    // for pod/service CRUD
+    podLister     kubelisters.PodLister
+    podControl    controller.PodControlInterface
+    serviceLister kubelisters.ServiceLister
+
+    // workQueue is a rate limited work queue. This is used to queue work to be
+    // processed instead of performing it as soon as a change happens. This
+    // means we can ensure we only process a fixed amount of resources at a
+    // time, and makes it easy to ensure we are never processing the same item
+    // simultaneously in two different workers.
+    workQueue workqueue.RateLimitingInterface
+
+    // recorder is an event recorder for recording Event resources to the
+    // Kubernetes API.
+    recorder record.EventRecorder
+
+    // A TTLCache of pod creates/deletes each TFReplica expects to see
+    expectations controller.ControllerExpectationsInterface
+}
+```
+
+### Distributed TensorFlow Configuration
+
+**Auto-Generated Arguments:**
+
+To make distributed TensorFlow work, user **MUST** implement a parser in their code like `ArgumentParser` to get the distributed TensorFlow configurations which generated by `tf-operator`, the common built-in arguments are:
+
+- `ps_hosts`: comma-separated list of hostname:port pairs.
+  + For example, `ps0.example.com:2222,ps1.example.com:2222`
+- `worker_hosts`: comma-separated list of hostname:port pairs.
+  + For example, `worker0.example.com:2222,worker1.example.com:2222,worker2.example.com:2222`
+- `job_name`: job name: worker or ps.
+- `task_index`: task index, should be >= 0.
+  + For example, worker task_index = 0 is the `chief` worker task the performs the variable initialization.
```

Comment:
I think it would be useful for these configurations to be env-dependant. For example we could run same job in local (notebook) single node for smoke testing and first metrics and then, ideally without code change, run it distributed to do actual learning.

Reply:
Thanks for your suggestion, and I think we already implement it since the Kubernetes could schedule all tasks of the job in one node if we only have one node.